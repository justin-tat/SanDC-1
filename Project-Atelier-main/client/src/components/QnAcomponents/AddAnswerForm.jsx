import React from 'react';
import axios from 'axios';
import AnswerPhotoUpload from './AnswerPhotoUpload.jsx';

class AddAnswerForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      answerBody: '',
      nickname: '',
      email: '',
      photos: [],
      isWarningShown: false
    };
    this.handleInputChange = this.handleInputChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.handlePhotos = this.handlePhotos.bind(this);
  }

  handleInputChange(event) {
    const target = event.target;
    const value = target.value;
    const name = target.name;
    this.setState({
      [name]: value
    });
  }

  handlePhotos(photos) {
    console.log('triggered handle photos');
    this.setState({
      photos: photos
    });
  }

  handleSubmit(event) {
    event.preventDefault();
    var questionId = this.props.questionId;
    var body = this.state.answerBody;
    var nickname = this.state.nickname;
    var email = this.state.email;
    var photos = this.state.photos;
    var validationResult = this.handleValidation(body, nickname, email);
    var productId = this.props.productId;

    if (validationResult) {
      this.setState({
        isWarningShown: false
      });
      this.props.addNewAnswer(questionId, body, nickname, email, photos, productId);
      this.props.closeAnswer();


    } else {
      this.setState({
        isWarningShown: true
      });
    }
  }

  handleValidation(answer, nickname, email) {
    if (answer.length === 0) {
      return false;
    } else if (nickname.length === 0) {
      return false;
    } else if (!(/^\w+([\.-]?\w+)*@\w+([\.-]?\w+)*(\.\w{2,3})+$/.test(email))) {
      return false;
    } else {
      return true;
    }
  }

  render() {
    let warning;
    if (this.state.isWarningShown) {
      warning = <div className='qna-add-answer-form-warning-message'>You didn't filled all the fields properly</div>;
    } else {
      warning = <div></div>;
    }
    return (
      <div className='qna-add-answer-form'>
        <div className='qna-add-answer-form-wrapper'>
          <div className='qna-add-answer-form-subtitle'>Submit your answer</div>
          <div className='qna-add-answer-form-product-name'>{this.props.productName}: {this.props.question_body}</div>
          <form>
            <label>
              <div className='qna-add-answer-form-your-answer'>
              Your answer*
              </div>
              <input
                className='qna-add-answer-form-input-answer'
                name='answerBody'
                type='text'
                maxLength='1000'
                value={this.state.answerBody}
                onChange={this.handleInputChange}
              />
            </label>
            <br />
            <label>
              <div className='qna-add-answer-form-your-nickname'>
              What's your nickname?*
              </div>
              <input
                className='qna-add-answer-form-input-nickname'
                name='nickname'
                type='text'
                maxLength='60'
                placeholder='Example: jack543'
                value={this.state.nickname}
                onChange={this.handleInputChange}
              />
              <br />
              <div className='qna-add-answer-form-warning-1'>
              For privacy reasons, do not use your full name or email address
              </div>
            </label>
            <br />
            <label>
              <div className='qna-add-answer-form-your-email'>
              Your email?*
              </div>
              <input
                className='qna-add-answer-form-input-email'
                name='email'
                type='text'
                maxLength='60'
                placeholder='Example: jack@email.com'
                value={this.state.email}
                onChange={this.handleInputChange}
              />
              <br />
              <div className='qna-add-answer-form-warning-2'>
             For authentication reasons, you will not be emailed
              </div>
            </label>
            <br />
            <label>
              <br />
              <AnswerPhotoUpload handlePhotos={this.handlePhotos} />
            </label>
            {warning}
            <input type="submit" className='qna-add-answer-form-submit' value="Submit" onClick = {this.handleSubmit} />
            <button type='button' className='qna-add-answer-form-close' onClick={this.props.closeAnswer}>Close without adding</button>
          </form>
        </div>
      </div>
    );
  }
}

export default AddAnswerForm;