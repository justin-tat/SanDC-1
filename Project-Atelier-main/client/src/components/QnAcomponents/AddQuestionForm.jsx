import React from 'react';
import axios from 'axios';

class AddQuestionForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      isValid: false,
      questionBody: '',
      nickname: '',
      email: '',
      isWarningShown: false

    };
    this.handleInputChange = this.handleInputChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleValidation = this.handleValidation.bind(this);
  }

  handleInputChange(event) {
    const target = event.target;
    const value = target.value;
    const name = target.name;
    this.setState({
      [name]: value
    });
  }

  handleSubmit(event) {
    event.preventDefault();
    var productId = this.props.productId;
    var body = this.state.questionBody;
    var nickname = this.state.nickname;
    var email = this.state.email;
    var validationResult = this.handleValidation(body, nickname, email);

    if (validationResult) {
      this.setState({
        isValid: true,
        isWarningShown: false
      });
      this.props.addQuestion(productId, body, nickname, email);
      this.props.closeForm();

    } else {
      this.setState({
        isWarningShown: true
      });
    }
  }
  handleValidation(question, nick, email) {
    if (question.length === 0) {
      return false;
    } else if (nick.length === 0) {
      return false;
    } else if (!(/^\w+([\.-]?\w+)*@\w+([\.-]?\w+)*(\.\w{2,3})+$/.test(email))) {
      return false;
    } else {
      return true;
    }
  }


  render() {
    let warning;
    if (this.state.isWarningShown) {
      warning = <div className='qna-add-question-form-warning-message'>You didn't properly filled all the fields</div>;
    } else {
      warning = <div></div>;
    }
    return (
      <div className='qna-add-new-question-form'>
        <div className='qna-add-question-form-wrapper'>
          {/* <div className = 'qna-add-question-main-title'>Ask a question</div> */}
          <div className ='qna-add-question-subtitle'>Ask a question about the {this.props.name}</div>
          <form onSubmit={this.handleSubmit}>
            <label>
              <div className ='qna-add-question-form-your-question'>Your question*</div>
              <input
                className='qna-add-question-input-questionbody'
                name='questionBody'
                id='questionBody'
                type='text'
                maxLength='1000'
                value={this.state.questionBody}
                onChange={this.handleInputChange}
              />
            </label>
            <br />
            <label>
              <div className='qna-add-question-form-your-nickname'>What's your nickname?*</div>
              <input
                className='qna-add-question-input-nickname'
                name='nickname'
                type='text'
                maxLength='60'
                placeholder='Example: jackson11'
                value={this.state.nickname}
                onChange={this.handleInputChange}
              />
              <br />
              <div className='qna-add-question-form-warning-1'>
               For privacy reasons, do not use your full name or email address
              </div>
            </label>
            <br />
            <label>
              <div className='qna-add-question-form-your-email'>
              Your email?*
              </div>
              <input
                className='qna-add-question-form-input-email'
                name='email'
                type='text'
                maxLength='60'
                placeholder='Example: sample@email.com'
                value={this.state.email}
                onChange={this.handleInputChange}
              />
              <br />
              <div className='qna-add-question-form-warning-2'>
               For authentication reasons, you will not be emailed
              </div>
            </label>
            <br />
            {warning}
            <input type="submit" className="qna-add-question-form-submit" value="Submit" onClick = {(e)=>this.handleSubmit(e)} />
            <button className='qna-add-question-form-close' onClick={()=> this.props.closeForm()}>Close without submitting</button>
          </form>
        </div>
      </div>

    );
  }
}

export default AddQuestionForm;