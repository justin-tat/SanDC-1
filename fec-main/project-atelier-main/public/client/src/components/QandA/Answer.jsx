import React from 'react';
import axios from 'axios';

class Answer extends React.Component{
  constructor(props){
    super(props)
    this.state = {
      haveData: false,
      allADisplayed: false
    }
    this.getMore = this.getMore.bind(this);
    this.getLess = this.getLess.bind(this);
    this.markHelpful = this.markHelpful.bind(this);
    this.report = this.report.bind(this);
    this.wasMarked = this.wasMarked.bind(this);
  }

  getMore (e) {
    this.props.clickTracker('Get more answers button', 'Answer.jsx');
    this.setState({ allADisplayed: true })
  }

  getLess (e) {
    this.props.clickTracker('collapse anwsers button', 'Answer.jsx');
    this.setState({ allADisplayed: false })
  }

  componentDidMount () {
    let id = this.props.props.question_id;
    this.setState({ questionId: id })
    this.getAnswerData(id);
  }

  dateFormat (date) {
    let month = date.slice(5, 7);
    if (month[0] === '0') {
      month = month[1];
    }
    let day = date.slice(8, 10);
    let year = date.slice(0, 4);
    let months = ['shiftingToMatch', 'January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December']

    return months[month] + ' ' + day + ', ' + year;
  }

  getAnswerData(id) {
    axios.post(`/getAnswerData`, {data: {id: id}})
    .then((result) => {
      this.setState({
        answerData: result.data.results,
        haveData: true
      })
    })
    .catch((error) => {
      console.error(error);
    })
  }

  markHelpful (e) {
    this.props.clickTracker('Mark Answer Helpful', 'Answer.jsx');
    let id = e.target.id;
    if (this.state[id] !== true) {
      axios('/markAHelpful', {data: {id: id}})
      .then((result) => {
        this.wasMarked(e);
        this.getAnswerData(this.props.props.question_id);
      })
      .catch((error) => {
        console.error(error);
      })
    }
  }

  wasMarked (e) {
    this.setState({ [e.target.id]: true })
  }

  report (e) {
    this.props.clickTracker('Report Answer Button', 'Answer.jsx');
    let id = e.target.id;
    axios.post('/reportAnswer', {data: {id: id}})
    .then((result) => {
      e.target.className = 'reported'
      e.target.innerText = 'Reported'
    })
    .catch((error) => {
      console.error(error);
    })
  }

  render() {
    if (!this.state.haveData) {
      return (
        <div className='loading'>
          Answers are loading
        </div>
      )
    } else if (!this.state.allADisplayed) {
      let eachAnwser = this.state.answerData.slice(0, 2).map((item) => {
        return (
          <span key={item.answer_id} className='answer'>
            {item.body} <br></br>
            by {item.answerer_name === 'Seller' ? <span className='seller'>{item.answerer_name}</span> : <span className='answerer'>{item.answerer_name}</span>}, {this.dateFormat(item.date)} <span className='helpful'>Helpful? <span className='yes' id={item.answer_id} onClick={this.markHelpful}>Yes ({item.helpfulness})</span> | <span className='report' id={item.answer_id} onClick={this.report}>Report</span></span><br></br>
          </span>
        )
      })
      return (
        <div className='answers'>
          <span className='aHeading'>A:</span> {eachAnwser}
          {this.state.answerData.length > 2 && <span className='loadMoreAnswers' onClick={this.getMore}>Load More Answers</span>}<br></br><br></br>
        </div>
      )
    } else {
      let eachAnwser = this.state.answerData.map((item) => {
        return (
          <div key={item.answer_id} className='answer'>
            {item.body} <br></br>
            by {item.answerer_name === 'Seller' ? <span className='seller'>{item.answerer_name}</span> : <span className='answerer'>{item.answerer_name}</span>}, {this.dateFormat(item.date)} <span className='helpful'>Helpful? <span className='yes' id={item.answer_id} onClick={this.markHelpful}>Yes ({item.helpfulness})</span> | <span className='report' id={item.answer_id} onClick={this.report}>Report</span></span><br></br>
          </div>
        )
      })
      return (
        <div className='answers'>
          <span className='aHeading'>A:</span>{eachAnwser} <br></br>
          <span onClick={this.getLess}>Collapse Answers</span><br></br>
        </div>
      )
    }
  }

}

export default Answer;