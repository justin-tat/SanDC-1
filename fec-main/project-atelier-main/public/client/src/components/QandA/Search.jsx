import React from 'react';

const Search = (props) => {
  //console.log('here', props);
  //let clicktracker = props.clickTracker('Search Bar', 'Search.jsx');
  return (
    <form onChange={props.searchBarChange} >
      <input type='text' id='search' size='50' placeholder='HAVE A QUESTION? SEARCH FOR ANSWERS...'></input>
    </form>
  )
};

export default Search;