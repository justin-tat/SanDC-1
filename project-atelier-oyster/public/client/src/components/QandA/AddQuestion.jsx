import React from 'react';
import axios from 'axios';

class AddQuestion extends React.Component {
  constructor(props) {
  super(props)
  this.state = {
    show: false
  };
  this.showModal = this.showModal.bind(this);
  this.hideModal = this.hideModal.bind(this);
  this.addQuestion = this.addQuestion.bind(this);
  }

  showModal (e) {
    this.props.clickTracker('Add Question Button', 'AddQuestion.jsx');
    this.setState({ show: true })
  }

  hideModal (e) {
    this.props.clickTracker('Submit Question Button', 'AddQuestion.jsx');
    this.setState({ show: false })
  }

  addQuestion (e) {
    e.preventDefault();
    let body = e.target[0].value;
    let name = e.target[1].value;
    let email = e.target[2].value;
    let id = this.props.id;

    axios.post(`/addQuestion`, {data: {body: body, name: name, email: email, product_id: id}})
    .then((result) => {
      console.log('Question was added');
      this.props.update(id, 1, 100);
    })
    .catch((error) => {
      console.error(error);
    })
    this.hideModal();
  }

  render () {
    if (this.state.show === false) {
      return(
        <span className='addQuestion stuff' onClick={this.showModal}>Add A Question +</span>
      )
    } else {
      return(
        <div className='modalBackground'>
          <div className='modalContainer'>
            <h3>Add Your Question</h3>
            <form className='newQForm' onSubmit={this.addQuestion}>
              <label className='bodylabel' for='body'>Your Question*</label>
              <input className='bodyInput' type='text' placeholder='Your Question Here' ref='body' name='body' required></input><br></br>
              <label className='nameLabel' for='name'>What is your nickname*</label>
              <input className='nameInput' type='text' placeholder='Example: jackson11!' ref='name' name='name' required></input><br></br>
              <div className='nameWarning'>For privacy reasons, do not use your full name or email address</div>
              <label className='emailLabel' for='email'>Your email*</label>
              <input className='emailInput' type='email' placeholder='Why did you like the product of not?' ref='email' name='email' required></input><br></br>
              <div className='emailWarning'>For authentication reasons, you will not be emailed<br></br>Fields marked with * are erquired</div>
              <input type='submit' value='submit'></input>
            </form>
          </div>
        </div>
      )
    }
  };
}

export default AddQuestion;