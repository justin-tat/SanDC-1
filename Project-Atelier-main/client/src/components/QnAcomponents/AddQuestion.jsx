import React from 'react';
import AddQuestionForm from './AddQuestionForm.jsx';

class AddQuestion extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      isModalShown: false,
      isAddButtonShown: true,
      productId: this.props.productId
    };
    this.clickOnAddQuestion = this.clickOnAddQuestion.bind(this);
    this.closeForm = this.closeForm.bind(this);

  }

  clickOnAddQuestion() {
    this.setState({
      isModalShown: true,
      isAddButtonShown: false
    });
    this.props.checkForm();
  }

  closeForm() {
    console.log('close form is triggered');
    this.setState({
      isModalShown: false,
      isAddButtonShown: true
    });
  }


  render() {
    let qnaAskQuestionModal,
      questionButton;
    if (this.state.isModalShown) {
      qnaAskQuestionModal = 'qna-ask-question-modal-shown';
    } else {
      qnaAskQuestionModal = 'qna-ask-question-modal-hidden';
    }
    // let modal,

    // if (this.state.isModalShown) {
    //   modal = <div></div>;
    // } else {
    //   modal = <div></div>;
    // }
    if (this.state.isAddButtonShown) {
      questionButton = <button className='qna-add-question-button' onClick={()=>this.clickOnAddQuestion()}>ADD A QUESTION</button>;
    } else {
      questionButton = <div></div>;
    }
    return (
      <div className='add-question-parent'>
        {questionButton}
        <div className={qnaAskQuestionModal}><AddQuestionForm
          name={this.props.name}
          productId={this.props.productId}
          addQuestion={this.props.addQuestion}
          closeForm={this.closeForm}
        />
        </div>
      </div>
    );
  }
}

export default AddQuestion;