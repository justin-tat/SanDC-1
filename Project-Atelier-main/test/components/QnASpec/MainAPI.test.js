import React from 'react';
import Enzyme from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';

Enzyme.configure({ adapter: new Adapter() });


import MainQnA from '../../../client/src/components/QnAcomponents/mainQnA.jsx';
import QuestionItem from '../../../client/src/components/QnAcomponents/QuestionsListItem.jsx';
import Answer from '../../../client/src/components/QnAcomponents/QuestionsListItemAnswer.jsx';
import AddAnswerForm from '../../../client/src/components/QnAcomponents/AddAnswerForm.jsx';
import AddQuestionForm from '../../../client/src/components/QnAcomponents/AddQuestionForm.jsx';
import Search from '../../../client/src/components/QnAcomponents/SearchQuestions.jsx';
import exampleQuestions from '../../../example/questions.js';
import exampleProducts from '../../../example/products.js';
import AddQuestion from '../../../client/src/components/QnAcomponents/AddQuestion.jsx';
import ClickData from '../../../client/src/components/ClickDataAnalytics.jsx';



import { mount, shallow } from 'enzyme';
import axios from 'axios';
import { act } from 'react-dom/test-utils';

import dummydata from './dummydata.js';
//import MockAdapter from 'axios-mock-adapter';

let example = {
  data: {
    results: [{
      'question_id': 37,
      'question_body': 'Why is this product cheaper here than other sites?',
      'question_date': '2018-10-18T00:00:00.000Z',
      'asker_name': 'williamsmith',
      'question_helpfulness': 4,
      'reported': false,
      'answers': {}
    }]
  }
};


describe('API calls in Main component', () => {
  afterEach(() => {
    jest.restoreAllMocks();
    jest.resetAllMocks();
  });
  it('mocks get API call', async () => {

    const response = { data: { name: 'mocked name' } };

    jest.spyOn(axios, 'get').mockResolvedValueOnce(response);

    const wrapper = shallow(<MainQnA currentProduct={exampleProducts.products[0]} productId={42} questionList={exampleQuestions.results}/>).dive();
    await act(async () => {
      await new Promise((resolve) => setTimeout(resolve, 0));
    });
    console.log(wrapper.state());
    expect (wrapper.state().productName).toEqual('mocked name');
    expect(axios.get).toHaveBeenCalledTimes(2);

  });


  it('mocks all get API call', async () => {
    const response1 = { data: { name: 'mocked name' } };
    const response2 = example;

    jest.spyOn(axios, 'get').mockResolvedValueOnce(response1)
      .mockResolvedValueOnce(response2);


    const wrapper = shallow(<MainQnA currentProduct={exampleProducts.products[0]} productId={42} questionList={exampleQuestions.results}/>).dive();

    await act(async () => {
      await new Promise((resolve) => setTimeout(resolve, 0));
    });

    expect (wrapper.state().productName).toEqual('mocked name');
    expect (wrapper.state().questions.length).toEqual(1);
    expect (wrapper.state().questions[0].question_id).toEqual(37);
    //expect(axios.get).toHaveBeenCalledWith('bla');
    expect(wrapper.instance()._isMounted).toEqual(true);
    expect(wrapper.state().isMoreQuestionsButtonShown).toEqual(false);

  });
  it('changes state then component is mounted', async () => {
    const response1 = { data: { name: 'mocked name' } };
    const response2 = example;

    jest.spyOn(axios, 'get').mockResolvedValueOnce(response1)
      .mockResolvedValueOnce(response2);

    const wrapper = shallow(<MainQnA currentProduct={exampleProducts.products[0]} productId={42} questionList={exampleQuestions.results}/>).dive();


    await act(async () => {
      await new Promise((resolve) => setTimeout(resolve, 0));
    });

    //
    console.log(wrapper.state());
    expect(wrapper.state().productName).toEqual('mocked name');

  });

  it('mocks put API call for clicking on helpful question', async () => {
    //parent component
    const response1 = { data: { name: 'mocked name' } };
    const response2 = example;

    jest.spyOn(axios, 'get').mockResolvedValueOnce(response1)
      .mockResolvedValueOnce(response2);

    //const wrapper = shallow(<MainQnA currentProduct={exampleProducts.products[0]} productId={42} questionList={exampleQuestions.results}/>).dive();

    const putResponse = example;

    jest.spyOn(axios, 'put'). mockResolvedValueOnce(putResponse);

    const wrapper = shallow(<MainQnA productId={42} currentProduct={exampleProducts.products[0]} questionsList={exampleQuestions.questions}/>).dive();

    const parentSpy = jest.spyOn(wrapper.instance(), 'clickOnHelpfulQuestion');
    //child component
    let spy = jest.fn();

    const component = mount(<QuestionItem clickOnHelpful={parentSpy}
      question={exampleQuestions.questions.results[0]} />);

    component.find('.qna-question-item-yes-button').simulate('click');

    expect(parentSpy).toHaveBeenCalled();
    expect(axios.put).toHaveBeenCalled();
    expect(wrapper.instance()._isMounted).toEqual(true);
    expect(axios.put).toHaveBeenCalledWith('/qna/updateQuestionHelp', {'params': {'productId': 1, 'questionId': 37}});

  });

  it('mocks put API call for clicking on helpful answer', async () => {
    //parent component
    const response1 = { data: { name: 'mocked name' } };
    const response2 = example;

    jest.spyOn(axios, 'get').mockResolvedValueOnce(response1)
      .mockResolvedValueOnce(response2);
    const putResponse = example;

    jest.spyOn(axios, 'put'). mockResolvedValueOnce(putResponse);

    const wrapper = shallow(<MainQnA productId={42} currentProduct={exampleProducts.products[0]} questionsList={exampleQuestions.questions}/>).dive();
    const parentSpy = jest.spyOn(wrapper.instance(), 'clickOnHelpfulAnswer');

    //child component
    let spy = jest.fn();

    const component = mount(<Answer clickOnHelpfulAnswer={parentSpy}
      answer={exampleQuestions.answers.results[0]} />);

    component.find('.qna-answer-item-yes-button').simulate('click');

    expect(parentSpy).toHaveBeenCalled();
    expect(axios.put).toHaveBeenCalled();
    expect(wrapper.instance()._isMounted).toEqual(true);
    expect(axios.put).toHaveBeenCalledWith('/qna/updateAnswerHelp', {'params': {'answerId': undefined, 'productId': 1}});

  });

  it('mocks put API call for clicking on report answer', async () => {

    //parent component
    const response1 = { data: { name: 'mocked name' } };
    const response2 = example;

    jest.spyOn(axios, 'get').mockResolvedValueOnce(response1)
      .mockResolvedValueOnce(response2);

    const putResponse = example;

    jest.spyOn(axios, 'put'). mockResolvedValueOnce(putResponse);

    const wrapper = shallow(<MainQnA productId={42} currentProduct={exampleProducts.products[0]} questionsList={exampleQuestions.questions}/>).dive();

    const parentSpy = jest.spyOn(wrapper.instance(), 'reportAnswer');

    //child component

    const component = mount(<Answer reportAnswer={parentSpy}
      answer={exampleQuestions.answers.results[0]} />);

    component.find('.qna-answer-item-report-button').simulate('click');

    expect(parentSpy).toHaveBeenCalled();
    expect(axios.put).toHaveBeenCalled();
    expect(wrapper.instance()._isMounted).toEqual(true);
    expect(axios.put).toHaveBeenCalledWith('/qna/reportAnswer', {'params': {'answerId': undefined, 'productId': 1}});
  });

  it('mocks post API call for adding a new answer', async () => {
    //parent component
    const response1 = { data: { name: 'mocked name' } };
    const response2 = example;

    jest.spyOn(axios, 'get').mockResolvedValueOnce(response1)
      .mockResolvedValueOnce(response2);

    const postResponse = example;

    jest.spyOn(axios, 'post'). mockResolvedValueOnce(postResponse);

    const parent = shallow(<MainQnA productId={42} currentProduct={exampleProducts.products[0]} questionsList={exampleQuestions.questions}/>).dive();
    const parentSpy = jest.spyOn(parent.instance(), 'addNewAnswer');

    //child component

    const formEventMocked = { preventDefault: jest.fn() };

    const passedMock1 = jest.fn();
    const passedMock2 = jest.fn();
    const wrapper = shallow(<AddAnswerForm addNewAnswer={parentSpy} closeAnswer={passedMock2} />);
    const spy1 = jest.spyOn(wrapper.instance(), 'handleInputChange');
    const spy2 = jest.spyOn(wrapper.instance(), 'handleSubmit');
    const spy3 = jest.spyOn(wrapper.instance(), 'handleValidation');

    const state = {
      answerBody: 'This is a answer from tests',
      nickname: 'gandalf',
      email: 'example@email.com'
    };
    wrapper.setState(state);

    wrapper.find('[type="submit"]').simulate('click', formEventMocked);
    expect(spy3).toBeCalledTimes(1);
    //expect(passedMock1).toBeCalledTimes(1);
    expect(passedMock2).toBeCalledTimes(1);
    expect(parentSpy).toHaveBeenCalled();
    expect(axios.post).toHaveBeenCalled();
    expect(parent.instance()._isMounted).toEqual(true);

  });

  it('mocks post API call for adding a new question', async () => {
    const response1 = { data: { name: 'mocked name' } };
    const response2 = example;

    jest.spyOn(axios, 'get').mockResolvedValueOnce(response1)
      .mockResolvedValueOnce(response2);

    //parent component

    const postResponse = example;

    jest.spyOn(axios, 'post'). mockResolvedValueOnce(postResponse);

    const parent = shallow(<MainQnA productId={42} currentProduct={exampleProducts.products[0]} questionsList={exampleQuestions.questions}/>).dive();
    const parentSpy = jest.spyOn(parent.instance(), 'addNewQuestion');
    const parentSpy2 = jest.spyOn(parent.instance(), 'updateQuestionList');

    //child component
    const formEventMocked = { preventDefault: jest.fn() };

    const passedMock1 = jest.fn();
    const passedMock2 = jest.fn();
    const wrapper = shallow(<AddQuestionForm addQuestion={parentSpy} closeForm={passedMock2} />);
    const spy1 = jest.spyOn(wrapper.instance(), 'handleInputChange');
    const spy2 = jest.spyOn(wrapper.instance(), 'handleSubmit');
    const spy3 = jest.spyOn(wrapper.instance(), 'handleValidation');

    const state = {
      questionBody: 'This is a question from tests',
      nickname: 'gandalf',
      email: 'example@email.com'
    };
    wrapper.setState(state);
    expect(wrapper).toMatchSnapshot();

    wrapper.find('[type="submit"]').simulate('click', formEventMocked);
    expect(wrapper.state().isValid).toEqual(true);
    expect(spy3).toBeCalledTimes(1);
    expect(passedMock2).toBeCalledTimes(1);
    expect(parentSpy).toHaveBeenCalled();
    expect(parent.instance()._isMounted).toEqual(true);
    expect(axios.post).toHaveBeenCalled();

  });

  it('mocks get API call for search', async () => {
    const response1 = { data: { name: 'mocked name' } };
    const response2 = example;

    jest.spyOn(axios, 'get').mockResolvedValueOnce(response1)
      .mockResolvedValueOnce(response2);

    //parent component
    const response = example;

    jest.spyOn(axios, 'get').mockResolvedValueOnce(response);


    const parent = shallow(<MainQnA productId={42} currentProduct={exampleProducts.products[0]} questionsList={exampleQuestions.questions.results}/>).dive();
    const parentSpy = jest.spyOn(parent.instance(), 'search');

    await act(async () => {
      await new Promise((resolve) => setTimeout(resolve, 0));
    });

    //child component

    const component = mount(<Search search={parentSpy} value='bla' />);
    const input = component.find('input');
    const event = {
      preventDefault() {},
      target: { value: 'bla' }
    };
    component.find('input').simulate('change', event);
    // expect(onSearchMock).toHaveBeenCalled();
    // expect(onSearchMock).toBeCalledWith('bla', true);
    //expect(parentSpy).toHaveBeenCalled();
    expect(axios.get).toHaveBeenCalledTimes(3);
    expect(parentSpy).toHaveBeenCalledWith('bla', true);
    //console.log('278', parent.state());
    expect(parent.instance()._isMounted).toEqual(true);

    //expect(parent.state().isMoreQuestionsButtonShown).toEqual(true);

  });

  xit('deals with error during search', async () => {

     const getResponse1 = { data: { name: 'mocked name' } };
    const getResponse2 = example;

    // jest.spyOn(axios, 'get').mockResolvedValueOnce(response1)
    //   .mockResolvedValueOnce(response2);

    //parent component

    const error = new Error('error');
    jest.spyOn(axios, 'get').mockResolvedValueOnce(getResponse1)
      .mockResolvedValueOnce(getResponse2)
      .mockRejectedValue(error);

    const parent = mount(<MainQnA productId={42}/>);
    const parentSpy = jest.spyOn(parent.instance(), 'search');

    await act(async () => {
      await new Promise((resolve) => setTimeout(resolve, 0));
    });

  });


  it('doesn\'t trigger search API calls if the query length is less than 3 chars ', async () => {

    //parent component
    const response1 = { data: { name: 'mocked name' } };
    const response2 = example;

    jest.spyOn(axios, 'get').mockResolvedValueOnce(response1)
      .mockResolvedValueOnce(response2);
    const response = example;

    jest.spyOn(axios, 'get').mockResolvedValueOnce(response);


    const parent = shallow(<MainQnA productId={42} currentProduct={exampleProducts.products[0]} questionsList={exampleQuestions.questions.results}/>).dive();
    const parentSpy = jest.spyOn(parent.instance(), 'search');

    await act(async () => {
      await new Promise((resolve) => setTimeout(resolve, 0));
    });

    //child component

    const component = mount(<Search search={parentSpy} value='bla' />);
    const input = component.find('input');
    const event = {
      preventDefault() {},
      target: { value: '42' }
    };
    component.find('input').simulate('change', event);

    expect(parentSpy).toHaveBeenCalled();
    expect(parentSpy).toHaveBeenCalledWith('42', false);
    expect(axios.get).toHaveBeenCalledTimes(2);
    expect(parent.instance()._isMounted).toEqual(true);


  });

  it('correctly changes state of main if add new question form is clicked ', async () => {
    //parent component
    const response1 = { data: { name: 'mocked name' } };
    const response2 = example;

    jest.spyOn(axios, 'get').mockResolvedValueOnce(response1)
      .mockResolvedValueOnce(response2);
    const postResponse = example;

    jest.spyOn(axios, 'post'). mockResolvedValueOnce(postResponse);

    const parent = shallow(<MainQnA productId={42} currentProduct={exampleProducts.products[0]} questionsList={exampleQuestions.questions}/>).dive();
    const parentSpy = jest.spyOn(parent.instance(), 'checkAddingNewQuestion');


    await act(async () => {
      await new Promise((resolve) => setTimeout(resolve, 0));
    });
    expect(parent.state().isAddNewQuestionClicked).toEqual(false);
    //child component

    const mockCheckForm = jest.fn();
    const wrapper = shallow(<AddQuestion checkForm={parentSpy} />);
    const spy = jest.spyOn(wrapper.instance(), 'clickOnAddQuestion');
    wrapper.find('.qna-add-question-button').simulate('click');
    expect(spy).toHaveBeenCalled();
    expect(parentSpy).toHaveBeenCalledTimes(1);
    expect(parent.state().isAddNewQuestionClicked).toEqual(true);



  });











});
