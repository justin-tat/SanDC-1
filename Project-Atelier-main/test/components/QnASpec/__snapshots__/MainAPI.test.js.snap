// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`API calls in Main component mocks post API call for adding a new question 1`] = `ShallowWrapper {}`;
